config_parser_test.cc

```diff
@@ -1,11 +1,68 @@
 #include "gtest/gtest.h"
 #include "config_parser.h"
 
-TEST(NginxConfigParserTest, SimpleConfig) {
-  NginxConfigParser parser;
-  NginxConfig out_config;
+// foo bar;
+TEST(NginxConfigTest, ToString) {
+    NginxConfigStatement statement;
+    statement.tokens_.push_back("foo");
+    statement.tokens_.push_back("bar");
+    EXPECT_EQ(statement.ToString(0), "foo bar;\n");
+}
+
+class NginxStringConfigTest : public ::testing::Test {
+	protected:
+    bool ParseString(const std::string config_string) {
+        std::stringstream config_stream(config_string);
+        return parser_.Parse(&config_stream, &out_config_);
+    }
+    NginxConfigParser parser_;
+    NginxConfig out_config_;
+};
 
-  bool success = parser.Parse("example_config", &out_config);
+TEST_F(NginxStringConfigTest, SyntaxErrorConfig) {
+    // from examples in class
+    EXPECT_FALSE(ParseString("")) << "Error: failed to detect empty as invalid";
+    EXPECT_FALSE(ParseString("#comment")) << "Error: failed to detect that just comments is invalid";
+    EXPECT_FALSE(ParseString("foo bar")) << "Error: failed to detect missing semicolon.";
+    EXPECT_FALSE(ParseString("server {listen 80;")) << "Error: failed to detect missing '}'";
+    EXPECT_FALSE(ParseString("listen 80;}")) << "Error: failed to detect missing '{'";
+    EXPECT_FALSE(ParseString("server {listen 80}")) << "Error: failed to detect missing semicolon within braces";
+    EXPECT_FALSE(ParseString("server {\n\
+                                server{listen 80;}")) << "Error: failed to detect error in nested braces";
+}
+
+TEST_F(NginxStringConfigTest, ValidSyntaxConfig) {
+    // from examples class
+    EXPECT_TRUE(ParseString("foo bar;")) << "Error: failed to correctly parse valid string";
+    EXPECT_EQ(1, out_config_.statements_.size()) << "Error: tokens parsed into incorrect number of statements";
+    EXPECT_EQ("foo", out_config_.statements_.at(0)->tokens_.at(0)) << "Error: incorrectly tokenized statement";
+    EXPECT_TRUE(ParseString("foo bar; baz bop;")) << "Error: failed to detect one line, two statements";
+}
+
+TEST_F(NginxStringConfigTest, NestedConfig) {
+    // tests nested cases
+    EXPECT_TRUE(ParseString("#test\n\
+                              server {\n\
+                                server {listen 80;}\n\
+                                     }")) << "Error: failed to parse doubly nested braces + comment correctly";
```

Comment:
To help readability, it would be nice if the error message (`<< "Error: ...`) were moved to the next line. For example: 

```c++
TEST_F(NginxStringConfigTest, NestedConfig) {
    EXPECT_TRUE(left_value, \n
                              right_value \)
        << "Error: ..."
```

That way, the comment is harder to miss. :) This doesn't apply to the other unit tests, as they were all short one-liners.